include/El/core/AbstractMatrix/decl.hpp

```diff
@@ -0,0 +1,421 @@
+    virtual ~AbstractMatrix() = default;
+
+    ///@}
+    /** @name Queries */
+    ///@{
+
+    /** @brief Get the number of rows in the matrix. */
+    size_type Height() const EL_NO_EXCEPT;
+
+    /** @brief Get the number of columns in the matrix. */
+    size_type Width() const EL_NO_EXCEPT;
+
+    /** @brief Get the leading dimension of the memory.
+     *
+     *  It is assumed that the matrix memory is stored column-major;
+     *  this information is required for most BLAS interactions, as
+     *  well as numerous internal kernels.
+     */
+    size_type LDim() const EL_NO_EXCEPT;
+
+    /** @brief Get the size of the underlying memory. */
+    virtual size_type MemorySize() const EL_NO_EXCEPT = 0;
+
+    /** @brief Get the the length of the matrix diagonal.
+     *
+     *  @param[in] offset If positive, a column offset. If negative, a
+     *                    row offset.
+     *
+     *  @return The length of the diagonal of the submatrix
+     *          A[:, offset:] if offset > 0 or
+     *          A[-offset:, :] if offset <= 0.
+     */
+    size_type DiagonalLength(
+        size_type offset = size_type{0}) const EL_NO_EXCEPT;
+
+    /** @brief Test if the matrix is a view.
+     *
+     *  Views are considered to have fixed size. Smaller views may be
+     *  constructed by creating new views with smaller dimensions.
+     */
+    bool Viewing() const EL_NO_EXCEPT;
+
+    /** @brief Test if the matrix has a fixed size.
+     *
+     *  Any AbstractMatrix may have a fixed size. This will prevent
+     *  resizing. All AbstractMatrix objects that represent views have
+     *  fixed size by definition.
+     */
+    bool FixedSize() const EL_NO_EXCEPT;
+
+    /** @brief Test if the matrix is logically @c const.
+     *
+     *  If @c this->Locked(), certain methods that are not marked @c
+     *  const may throw with exceptions indicating that a logical @c
+     *  const condition would be violated.
+     */
+    bool Locked() const EL_NO_EXCEPT;
+
+    /** @brief Test if the matrix is empty (i.e., has size 0x0). */
+    bool IsEmpty() const EL_NO_EXCEPT;
+
+    /** @brief Test if the matrix memory is contiguous.
+     *
+     *  A matrix is considered contiguous if any of the following hold:
+     *    - The matrix height equals its leading dimension.
+     *    - The matrix has width 1.
+     *    - The matrix is empty.
+     */
+    bool Contiguous() const EL_NO_EXCEPT;
+
+    ///@}
+    /** @name Advanced queries */
+    ///@{
+
+    /** @brief Get the ViewType of the matrix. */
+    El::ViewType ViewType() const EL_NO_EXCEPT;
+
+    /** @brief Get the device type on which memory is resident. */
+    virtual Device GetDevice() const EL_NO_EXCEPT = 0;
+
+    /** @brief Get the underlying memory mode.
+     *
+     *  @todo (trb) Why is this exposed here? This is intimately tied
+     *  to device allocations and the like. I'm skeptical this is
+     *  required at this level of the interface.
```

Comment:
The third option is, of course, my favorite of these options. I would be extremely hesitant to expose the `Memory` object to users. They would have to cast that because there is not a common base class. Since the memory mode is just an `unsigned int` that might mean different things depending on which device you're on, I'm definitely a fan of forcing a cast to a concrete type so that users have a slightly higher probability of being sure they know what they're doing.